charmcraft/parts.py

```diff
@@ -0,0 +1,243 @@
+        """
+        plugin_data = plugins.extract_plugin_properties(
+            data, plugin_name="charm", required=["source"]
+        )
+        return cls(**plugin_data)
+
+
+class CharmPlugin(plugins.Plugin):
+    """Build the charm and prepare for packing.
+
+    The craft-parts charm plugin prepares the charm payload for packing. Common
+    plugin and source keywords can be used.
+
+    Extra files to be included in the charm payload must be listed under
+    the ``prime`` file filter.
+    """
+
+    properties_class = CharmPluginProperties
+
+    @classmethod
+    def get_build_snaps(cls) -> Set[str]:
+        """Return a set of required snaps to install in the build environment."""
+        return set()
+
+    def get_build_packages(self) -> Set[str]:
+        """Return a set of required packages to install in the build environment."""
+        return {
+            "python3-pip",
+            "python3-setuptools",
+            "python3-wheel",
+        }
+
+    def get_build_environment(self) -> Dict[str, str]:
+        """Return a dictionary with the environment to use in the build step."""
+        return {}
+
+    def get_build_commands(self) -> List[str]:
+        """Return a list of commands to run during the build step."""
+        build_env = dict(LANG="C.UTF-8", LC_ALL="C.UTF-8")
+        for key in [
+            "PATH",
+            "SNAP",
+            "SNAP_ARCH",
+            "SNAP_NAME",
+            "SNAP_VERSION",
+            "http_proxy",
+            "https_proxy",
+            "no_proxy",
+        ]:
+            if key in os.environ:
+                build_env[key] = os.environ[key]
+
+        env_flags = [f"{key}={value}" for key, value in build_env.items()]
+
+        # invoke the charm builder
+        build_cmd = [
+            "env",
+            "-i",
+            *env_flags,
+            sys.executable,
+            "-I",
+            charm_builder.__file__,
+            "--charmdir",
+            str(self._part_info.part_build_dir),
+            "--builddir",
+            str(self._part_info.part_install_dir),
+        ]
+
+        if self._part_info.entrypoint:
+            entrypoint = self._part_info.part_build_dir / self._part_info.entrypoint
+            build_cmd.extend(["--entrypoint", str(entrypoint)])
+
+        for req in self._part_info.requirements:
+            build_cmd.extend(["-r", req])
+
+        commands = [" ".join(shlex.quote(i) for i in build_cmd)]
+
+        return commands
+
+
+def setup_parts():
+    """Initialize craft-parts plugins."""
+    plugins.register({"charm": CharmPlugin})
+
+
+def validate_part(data: Dict[str, Any]) -> None:
+    """Validate the given part data against common and plugin models.
+
+    :param data: The part data to validate.
+    """
+    if not isinstance(data, dict):
+        raise TypeError("value must be a dictionary")
+
+    # copy the original data, we'll modify it
+    spec = data.copy()
+
+    plugin_name = spec.get("plugin", "")
```

Comment:
No need for the default `""`, right? if you get None a ValueError will be raised below anyway...